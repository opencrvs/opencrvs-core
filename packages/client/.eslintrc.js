/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
module.exports = {
  extends: [
    '../../.eslintrc.js',
    'eslint-config-react-app',
    'plugin:react/recommended'
  ],
  plugins: ['react', 'formatjs'],
  env: {
    es6: true,
    browser: true,
    node: true
  },
  rules: {
    'react-hooks/exhaustive-deps': 'error',
    '@typescript-eslint/no-explicit-any': 'warn',
    'no-restricted-imports': [
      'error',
      {
        paths: [
          {
            name: '@sentry/browser',
            message: `Errors should be sent to Sentry from few centralised places of our codebase.
Query component now sends errors automatically to Sentry.`
          },
          {
            name: '@apollo/client/react/components',
            importNames: ['Query'],
            message: `Please use our own <Query /> component instead from components/Query.tsx`
          },
          {
            name: 'date-fns',
            message: 'Please use submodules instead'
          },
          {
            name: '@opencrvs/commons',
            message:
              'Importing directly from `@opencrvs/commons` is not allowed. Use `@opencrvs/commons/client` instead.'
          }
        ],
        patterns: ['@opencrvs/commons/*', '!@opencrvs/commons/client']
      }
    ],
    'formatjs/enforce-id': 'error',
    'react/no-unescaped-entities': 'off',
    'react/destructuring-assignment': 'off',
    'react/jsx-filename-extension': [
      1,
      {
        extensions: ['.tsx']
      }
    ],
    'react/boolean-prop-naming': 'off',
    'react/sort-comp': 'off',
    'react/sort-prop-types': 'off',
    'react/prop-types': 'off'
  },
  settings: {
    react: {
      pragma: 'React',
      version: 'detect'
    }
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      parserOptions: {
        project: ['./tsconfig.json']
      }
    }
  ],

  globals: {},
  parser: '@typescript-eslint/parser'
}
