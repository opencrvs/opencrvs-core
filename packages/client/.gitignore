# See https://help.github.com/ignore-files/ for more about ignoring files.

# dependencies
/node_modules

# testing
/coverage

# production
/build

# misc
*/assets/locations.json
.DS_Store
.env.local
.env.development.local
.env.test.local
.env.production.local
/src/extract-translations.js
/dev-dist

npm-debug.log*
yarn-debug.log*
yarn-error.log*


src/i18n/locales/bn.csv
src/i18n/locales/en.csv

.vscode
graphql.schema.json
.lh
.idea

*storybook.log
storybook-static