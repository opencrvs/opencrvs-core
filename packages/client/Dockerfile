ARG  BRANCH=develop
FROM ghcr.io/opencrvs/ocrvs-base:${BRANCH}

USER node

COPY --chown=node:node packages/gateway /app/packages/gateway
COPY --chown=node:node packages/events /app/packages/events

WORKDIR /app/packages/components
COPY --chown=node:node packages/components /app/packages/components

RUN yarn install --frozen-lockfile  && yarn build

# Placeholders:
# Variables should be substituted in docker-compose/kubernetes manifest
ENV CONTENT_SECURITY_POLICY_WILDCARD="{{CONTENT_SECURITY_POLICY_WILDCARD}}"
ENV COUNTRY_CONFIG_URL_INTERNAL="{{COUNTRY_CONFIG_URL_INTERNAL}}"
ENV GATEWAY_URL_INTERNAL="{{GATEWAY_URL_INTERNAL}}"
WORKDIR /app/packages/client
COPY --chown=node:node packages/client /app/packages/client
RUN yarn install --frozen-lockfile && yarn build

FROM nginxinc/nginx-unprivileged:1.28-alpine

COPY --from=0 --chown=nginx:nginx /app/packages/client/build/ /usr/share/nginx/html/

COPY --chmod=555 packages/client/nginx-deploy-config.sh /docker-entrypoint.d/11-nginx-deploy-config.sh
COPY --chown=nginx:nginx packages/client/nginx.conf /etc/nginx/conf.d/default.conf
EXPOSE 3000
