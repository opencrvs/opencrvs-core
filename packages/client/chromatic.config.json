{
  "onlyChanged": true,
  "projectId": "Project:67862a3f4457d14c317e5cca",
  "storybookBaseDir": "packages/client",
  "zip": true
}
