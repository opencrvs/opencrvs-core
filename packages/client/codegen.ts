/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import type { CodegenConfig } from '@graphql-codegen/cli'

const config: CodegenConfig = {
  schema: [
    {
      'http://localhost:7070/graphql': {
        headers: { Authorization: `Bearer ${process.env.TOKEN}` }
      }
    }
  ],
  documents: ['src/**/*.ts*', '!src/utils/gateway.ts', '!src/**/*.test.ts*'],
  generates: {
    './src/utils/gateway.ts': {
      config: {
        scalars: {
          PlainDate: 'PlainDate'
        }
      },
      plugins: [
        'typescript',
        'typescript-operations',
        {
          add: {
            content: `/*
             * This Source Code Form is subject to the terms of the Mozilla Public
             * License, v. 2.0. If a copy of the MPL was not distributed with this
             * file, You can obtain one at https://mozilla.org/MPL/2.0/.
             *
             * OpenCRVS is also distributed under the terms of the Civil Registration
             * & Healthcare Disclaimer located at http://opencrvs.org/license.
             *
             * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
             */

            /*
             * DO NOT EDIT! This file is auto-generated by yarn generate-gateway-types - see 'codegen.yml'
             */

            /* eslint-disable */
            import { PlainDate } from '@client/utils/date-formatting'

            `
          }
        }
      ]
    },
    './graphql.schema.json': {
      plugins: ['introspection']
    }
  },
  overwrite: true
}
export default config
