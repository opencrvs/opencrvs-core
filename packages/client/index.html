<!--This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.

OpenCRVS is also distributed under the terms of the Civil Registration
& Healthcare Disclaimer located at http://opencrvs.org/license.

Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
-->

<!DOCTYPE html>
<html lang="en" translate="no">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1, shrink-to-fit=no"
    />
    <meta name="theme-color" content="#4c68c1" />

    <!--
    Inject configuration values from public/config.js into window
    replace this file to change runtime configuration
  -->
    <script src="/api/countryconfig/client-config.js"></script>
    <!--
      manifest.json provides metadata used when your web app is added to the
      homescreen on Android. See https://developers.google.com/web/fundamentals/engage-and-retain/web-app-manifest/
    -->
    <link rel="manifest" href="/manifest.json" />
    <link rel="shortcut icon" href="/favicon.ico" />
    <!--
      Unlike "/favicon.ico" or "favicon.ico", "/favicon.ico" will
      work correctly both with client-side routing and a non-root public URL.
      Learn how to configure a non-root public URL by running `npm run build`.
    -->
    <title>Loading...</title>

    <style>
      body,
      html {
        background-color: #f8f8f8;
        margin: 0;
      }
      .pgrs-wrapper1240 {
        height: 100vh;
        width: 100vw;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
        opacity: 0;
        animation: fadeIn 2s forwards;
        font-family: sans-serif;
      }
      .progress-bar {
        width: 276px;
        height: 8px;
        margin-top: 27px;
        border-radius: 100px;
        background: #d9d9d9;
      }
      #root-app-progress {
        width: 0;
        height: 8px;
        border-radius: 100px;
        background: #4972bb;
        animation: progressAnimation 200s ease-out forwards;
      }
      @keyframes progressAnimation {
        0%,
        1% {
          min-width: 0%;
        }
        3% {
          min-width: 80%;
        }
        6% {
          min-width: 91%;
        }
        10% {
          min-width: 94%;
        }
        30% {
          min-width: 96%;
        }
        100% {
          min-width: 99%;
        }
      }
      @keyframes fadeIn {
        0%,
        85% {
          opacity: 0;
        }
        100% {
          opacity: 1;
        }
      }
    </style>
  </head>
  <body>
    <noscript> You need to enable JavaScript to run this app. </noscript>
    <div id="root">
      <div class="pgrs-wrapper1240">
        <img
          src="./images/logo-90x90.svg"
          width="90"
          height="90"
          alt="OpenCRVS logo"
        />
        <div class="progress-bar">
          <div id="root-app-progress"></div>
        </div>
        <p>Installing application...</p>
      </div>
    </div>
    <script type="module" src="/src/index.tsx"></script>
  </body>
</html>
