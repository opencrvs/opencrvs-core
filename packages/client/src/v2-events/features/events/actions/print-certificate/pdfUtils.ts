/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import {
  IntlShape,
  MessageDescriptor,
  createIntl,
  createIntlCache
} from 'react-intl'
import Handlebars from 'handlebars'
import htmlToPdfmake from 'html-to-pdfmake'
import type {
  Content,
  TDocumentDefinitions,
  TFontFamilyTypes
} from 'pdfmake/interfaces'
import pdfMake from 'pdfmake/build/pdfmake'
import { Location } from '@events/service/locations/locations'
import { isEqual } from 'lodash'
import {
  EventState,
  User,
  LanguageConfig,
  FieldValue,
  EventConfig,
  getMixedPath,
  EventMetadata,
  EventStatus,
  DEFAULT_DATE_OF_EVENT_PROPERTY
} from '@opencrvs/commons/client'
import { DateField } from '@client/v2-events/features/events/registered-fields'
import { getHandlebarHelpers } from '@client/forms/handlebarHelpers'
import { isMobileDevice } from '@client/utils/commonUtils'
import { getUsersFullName } from '@client/v2-events/utils'
import { getFormDataStringifier } from '@client/v2-events/hooks/useFormDataStringifier'
import { LocationSearch } from '@client/v2-events/features/events/registered-fields'

interface FontFamilyTypes {
  normal: string
  bold: string
  italics: string
  bolditalics: string
}

type CertificateConfiguration = Partial<{
  fonts: Record<string, FontFamilyTypes>
}>

function findUserById(userId: string, users: User[]) {
  const user = users.find((u) => u.id === userId)

  if (!user) {
    return {
      name: '',
      signature: ''
    }
  }

  return {
    name: getUsersFullName(user.name, 'en'),
    signature: user.signatureFilename ?? ''
  }
}

export const stringifyEventMetadata = ({
  metadata,
  intl,
  locations,
  users
}: {
  metadata: NonNullable<EventMetadata & { modifiedAt: string }>
  intl: IntlShape
  locations: Location[]
  users: User[]
}) => {
  return {
    modifiedAt: DateField.stringify(intl, metadata.modifiedAt),
    assignedTo: findUserById(metadata.assignedTo ?? '', users),
    // @TODO: DATE_OF_EVENT config needs to be defined some other way and bake it in.
    dateOfEvent: metadata.dateOfEvent
      ? DateField.stringify(intl, metadata.dateOfEvent)
      : DateField.stringify(intl, metadata[DEFAULT_DATE_OF_EVENT_PROPERTY]),
    createdAt: DateField.stringify(intl, metadata.createdAt),
    createdBy: findUserById(metadata.createdBy, users),
    createdAtLocation: LocationSearch.stringify(
      intl,
      locations,
      metadata.createdAtLocation
    ),
    updatedAt: DateField.stringify(intl, metadata.updatedAt),
    updatedBy: metadata.updatedBy
      ? findUserById(metadata.updatedBy, users)
      : '',
    id: metadata.id,
    type: metadata.type,
    trackingId: metadata.trackingId,
    status: EventStatus.REGISTERED,
    updatedByUserRole: metadata.updatedByUserRole,
    updatedAtLocation: LocationSearch.stringify(
      intl,
      locations,
      metadata.updatedAtLocation
    ),
    flags: [],
    legalStatuses: {
      [EventStatus.DECLARED]: metadata.legalStatuses.DECLARED
        ? {
            createdAt: DateField.stringify(
              intl,
              metadata.legalStatuses.DECLARED.createdAt
            ),
            createdBy: findUserById(
              metadata.legalStatuses.DECLARED.createdBy,
              users
            ),
            createdAtLocation: LocationSearch.stringify(
              intl,
              locations,
              metadata.legalStatuses.DECLARED.createdAtLocation
            ),
            acceptedAt: DateField.stringify(
              intl,
              metadata.legalStatuses.DECLARED.acceptedAt
            ),
            createdByRole: metadata.legalStatuses.DECLARED.createdByRole,
            createdBySignature:
              metadata.legalStatuses.DECLARED.createdBySignature
          }
        : null,
      [EventStatus.REGISTERED]: metadata.legalStatuses.REGISTERED
        ? {
            createdAt: DateField.stringify(
              intl,
              metadata.legalStatuses.REGISTERED.createdAt
            ),
            createdBy: findUserById(
              metadata.legalStatuses.REGISTERED.createdBy,
              users
            ),
            createdAtLocation: LocationSearch.stringify(
              intl,
              locations,
              metadata.legalStatuses.REGISTERED.createdAtLocation
            ),
            acceptedAt: DateField.stringify(
              intl,
              metadata.legalStatuses.REGISTERED.acceptedAt
            ),
            createdByRole: metadata.legalStatuses.REGISTERED.createdByRole,
            registrationNumber:
              metadata.legalStatuses.REGISTERED.registrationNumber,
            createdBySignature: metadata.legalStatuses.REGISTERED
              .createdBySignature
              ? new URL(
                  metadata.legalStatuses.REGISTERED.createdBySignature,
                  window.config.MINIO_BASE_URL
                ).href
              : undefined
          }
        : null
    }
  }
}

const certificateBaseTemplate = {
  definition: {
    pageMargins: [0, 0, 0, 0] as [number, number, number, number],
    defaultStyle: {
      font: 'notosans'
    },
    content: []
  },
  fonts: {}
}

function isMessageDescriptor(obj: unknown): obj is MessageDescriptor {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    'id' in obj &&
    'defaultMessage' in obj &&
    typeof (obj as MessageDescriptor).id === 'string' &&
    typeof (obj as MessageDescriptor).defaultMessage === 'string'
  )
}

function formatAllNonStringValues(
  templateData: EventState,
  intl: IntlShape
): EventState {
  const formattedData: EventState = {}

  for (const key of Object.keys(templateData)) {
    const value = templateData[key]

    if (isMessageDescriptor(value)) {
      formattedData[key] = intl.formatMessage(value)
    } else if (Array.isArray(value)) {
      // Address field: country label is a MessageDescriptor but others are strings
      formattedData[key] = value
        .filter(Boolean)
        .map((item) =>
          isMessageDescriptor(item) ? intl.formatMessage(item) : item
        )
        .join(', ')
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    } else if (typeof value === 'object' && value !== null) {
      formattedData[key] = formatAllNonStringValues(
        value satisfies EventState,
        intl
      ) as FieldValue
    } else {
      formattedData[key] = String(value)
    }
  }

  return formattedData
}

const cache = createIntlCache()

export function compileSvg({
  templateString,
  $metadata,
  $declaration,
  locations,
  users,
  language,
  config
}: {
  templateString: string
  $metadata: EventMetadata & { modifiedAt: string }
  $declaration: EventState
  locations: Location[]
  users: User[]
  language: LanguageConfig
  config: EventConfig
}): string {
  const intl = createIntl(
    {
      locale: language.lang,
      messages: language.messages
    },
    cache
  )

  const customHelpers = getHandlebarHelpers()

  for (const helperName of Object.keys(customHelpers)) {
    /*
     * Note for anyone adding new context variables to handlebar helpers:
     * Everything you expose to country config's here will become API surface area,
     * This means that countries will become dependant on it and it will be hard to remove or rename later on.
     * If you need to expose the full record, please consider only exposing the specific values you know are needed.
     * Otherwise what happens is that we lose the ability to refactor and remove things later on.
     */
    const helper = customHelpers[helperName]({ intl })
    Handlebars.registerHelper(helperName, helper)
  }

  /**
   * Handlebars helper: $lookup
   *
   * Resolves a value from the given property path within the combined $state and $declaration objects. useful for extracting specific properties from complex structures like `child.address.other`
   * and optionally returns a nested field from the resolved value. e.g. when defaultValue is set to $user.province, it resolves to the matching id.
   *
   * @param obj - Object to look up. This is for providing the same interface as the handlebars 'lookup'. It is not used as is.
   *  @param propertyPath - $declaration or $state property to look up without the top-level property name.
   *  @returns - a nested field from the resolved value.
   *
   * @example {'foo.bar.baz': 'quix' } // $lookup 'foo.bar.baz' => 'quix'
   * @example {'foo': {'bar': {'baz': 'quix'}} } // $lookup 'foo.bar.baz' => 'quix'
   * @example { 'informant.address': { 'other': { 'district': 'quix' } } } // $lookup 'informant.address.other.district' => 'quix'
   */
  function $lookup(obj: EventMetadata | EventState, propertyPath: string) {
    const stringifyDeclaration = getFormDataStringifier(intl, locations)
    const fieldConfigs = config.declaration.pages.flatMap((x) => x.fields)
    const resolvedDeclaration = stringifyDeclaration(fieldConfigs, $declaration)

    const resolvedMetadata = stringifyEventMetadata({
      metadata: $metadata,
      intl,
      locations,
      users
    })

    if (isEqual($metadata, obj)) {
      return getMixedPath(resolvedMetadata, propertyPath)
    }
    return getMixedPath(resolvedDeclaration, propertyPath)
  }

  Handlebars.registerHelper('$lookup', $lookup)

  /**
   * Handlebars helper: $intl
   *
   * Usage example in SVG template:
   *   <tspan>{{ $intl 'constants' (lookup $declaration "child.gender") }}</tspan>
   *
   * This helper dynamically constructs a translation key by joining multiple string parts
   * (e.g., 'constants.male') and uses `intl.formatMessage` to fetch the localized translation.
   *
   * In the example above, `"child.gender"` resolves to a value like `"male"` which forms
   * part of the translation key: `constants.male`.
   *
   * - If any of the parts is undefined (e.g., gender not provided), it returns an empty string to prevent rendering issues.
   * - If the translation for the constructed ID is missing, it falls back to showing: 'Missing translation for [id]'.
   *
   * This is especially useful in templates where dynamic values (like gender, marital status, etc.)
   * need to be translated using i18n keys constructed from user-provided data.
   */
  Handlebars.registerHelper(
    '$intl',

    function (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this: any,
      ...args: [...(string | undefined)[], Handlebars.HelperOptions]
    ) {
      // If even one of the parts is undefined, then return empty string
      const idParts = args.slice(0, -1)
      if (idParts.some((part) => part === undefined)) {
        return ''
      }

      const id = idParts.join('.')

      return intl.formatMessage({
        id,
        defaultMessage: 'Missing translation for ' + id
      })
      /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    } as any /* This is here because Handlebars typing is insufficient and we can make the function type stricter */
  )

  /**
   * Handlebars helper: $OR
   * Returns the first truthy value between v1 and v2.
   */
  Handlebars.registerHelper(
    '$or',
    function (
      /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
      v1: any,
      /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
      v2: any
    ) {
      return !!v1 ? v1 : v2
    }
  )

  /**
   * Handlebars helper: ifCond
   *
   * Usage example in template:
   *   {{#ifCond value1 '===' value2}} ... {{/ifCond}}
   *
   * This helper compares two values (`v1` and `v2`) using the specified operator and
   * conditionally renders a block based on the result of the comparison.
   *
   * Supported operators:
   *   - '===' : strict equality
   *   - '!==' : strict inequality
   *   - '<', '<=', '>', '>=' : numeric/string comparisons
   *   - '&&' : both values must be truthy
   *   - '||' : at least one value must be truthy
   *
   * If the condition is met, it renders the main block (`options.fn(this)`),
   * otherwise it renders the `else` block (`options.inverse(this)`).
   *
   * This helper is useful for adding conditional logic directly within Handlebars templates.
   */
  Handlebars.registerHelper(
    'ifCond',
    function (
      /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
      this: any,
      v1: string,
      operator: string,
      v2: string,
      options: Handlebars.HelperOptions
    ) {
      switch (operator) {
        case '===':
          return v1 === v2 ? options.fn(this) : options.inverse(this)
        case '!==':
          return v1 !== v2 ? options.fn(this) : options.inverse(this)
        case '<':
          return v1 < v2 ? options.fn(this) : options.inverse(this)
        case '<=':
          return v1 <= v2 ? options.fn(this) : options.inverse(this)
        case '>':
          return v1 > v2 ? options.fn(this) : options.inverse(this)
        case '>=':
          return v1 >= v2 ? options.fn(this) : options.inverse(this)
        case '&&':
          return v1 && v2 ? options.fn(this) : options.inverse(this)
        case '||':
          return v1 || v2 ? options.fn(this) : options.inverse(this)
        default:
          return options.inverse(this)
      }
    }
  )

  const template = Handlebars.compile(templateString)
  $declaration = formatAllNonStringValues($declaration, intl)
  const data = {
    $declaration,
    $metadata,
    $references: {
      locations,
      users
    }
  }
  const output = template(data)
  return output
}

export function addFontsToSvg(
  svgString: string,
  fonts: Record<string, FontFamilyTypes>
) {
  const parser = new DOMParser()
  const doc = parser.parseFromString(svgString, 'image/svg+xml')
  const svg = doc.documentElement
  const style = document.createElement('style')
  style.innerHTML = Object.entries(fonts)
    .flatMap(([font, families]) =>
      Object.entries(families).map(
        ([family, url]) => `
@font-face {
font-family: "${font}";
font-weight: ${family};
src: url("${url}") format("truetype");
}`
      )
    )
    .join('')
  svg.prepend(style)
  const serializer = new XMLSerializer()
  return serializer.serializeToString(svg)
}

export function svgToPdfTemplate(
  svg: string,
  certificateFonts: CertificateConfiguration
) {
  const pdfTemplate: PdfTemplate = {
    ...certificateBaseTemplate,
    definition: {
      ...certificateBaseTemplate.definition,
      defaultStyle: {
        font:
          Object.keys(certificateFonts)[0] ||
          certificateBaseTemplate.definition.defaultStyle.font
      }
    },
    fonts: {
      ...certificateBaseTemplate.fonts,
      ...certificateFonts
    }
  }

  const parser = new DOMParser()
  const svgElement = parser.parseFromString(
    svg,
    'image/svg+xml'
  ).documentElement

  const widthValue = svgElement.getAttribute('width')
  const heightValue = svgElement.getAttribute('height')

  if (widthValue && heightValue) {
    const width = Number.parseInt(widthValue)
    const height = Number.parseInt(heightValue)
    pdfTemplate.definition.pageSize = {
      width,
      height
    }
    if (width > height) {
      pdfTemplate.definition.pageOrientation = 'landscape'
    }
  }

  const foreignObjects = svgElement.getElementsByTagName('foreignObject')
  const absolutelyPositionedHTMLs: Content[] = []
  for (const foreignObject of foreignObjects) {
    const width = Number.parseInt(foreignObject.getAttribute('width') ?? '0')
    const x = Number.parseInt(foreignObject.getAttribute('x') ?? '0')
    const y = Number.parseInt(foreignObject.getAttribute('y') ?? '0')
    const htmlContent = foreignObject.innerHTML
    const pdfmakeContent = htmlToPdfmake(htmlContent, {
      ignoreStyles: ['font-family']
    })
    absolutelyPositionedHTMLs.push({
      columns: [
        {
          width,
          stack: pdfmakeContent
        }
      ],
      absolutePosition: { x, y }
    } as Content)
  }

  pdfTemplate.definition.content = [
    {
      svg
    },
    ...absolutelyPositionedHTMLs
  ]

  return pdfTemplate
}

interface PdfTemplate {
  definition: TDocumentDefinitions
  fonts: Record<string, TFontFamilyTypes>
}

export function printAndDownloadPdf(
  template: PdfTemplate,
  declarationId: string
) {
  const pdf = pdfMake.createPdf(template.definition, undefined, template.fonts)
  if (isMobileDevice()) {
    pdf.download(`${declarationId}`)
  } else {
    pdf.print()
  }
}
