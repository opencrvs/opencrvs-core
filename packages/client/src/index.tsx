/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
// eslint-disable-next-line import/no-unassigned-import
import 'focus-visible/dist/focus-visible.js'
import * as actions from '@client/notification/actions'
import { storage } from '@client/storage'
import { createStore } from '@client/store'
import * as React from 'react'
import { createRoot } from 'react-dom/client'
// eslint-disable-next-line no-restricted-imports
import { SubmissionController } from '@client/SubmissionController'
import * as Sentry from '@sentry/react'
import { BrowserTracing } from '@sentry/tracing'
import { createBrowserRouter } from 'react-router-dom'
import WebFont from 'webfontloader'
import { App, routesConfig } from './App'
import { APPLICATION_VERSION } from './utils/constants'

WebFont.load({
  google: {
    families: ['Noto+Sans:600', 'Noto+Sans:500', 'Noto+Sans:400']
  }
})

storage.configStorage('OpenCRVS')

const { store } = createStore()

if (
  window.location.hostname !== 'localhost' &&
  window.location.hostname !== '127.0.0.1'
) {
  // setup error reporting using sentry
  if (window.config.SENTRY) {
    Sentry.init({
      release: APPLICATION_VERSION,
      environment: import.meta.env.NODE_ENV,
      integrations: [new BrowserTracing()],

      // We recommend adjusting this value in production, or using tracesSampler
      // for finer control
      tracesSampleRate: 1.0,
      dsn: window.config.SENTRY
    })
  }
}

function userReconnectedToast() {
  const action = actions.showUserReconnectedToast()
  store.dispatch(action)
}

window.addEventListener('online', userReconnectedToast)

const container = document.getElementById('root')
const root = createRoot(container!)

const router = createBrowserRouter(routesConfig)

root.render(<App router={router} store={store} />)

new SubmissionController(store).start()
