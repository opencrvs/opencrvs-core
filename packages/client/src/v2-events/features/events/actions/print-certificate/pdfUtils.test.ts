/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createIntl } from 'react-intl'
import createFetchMock from 'vitest-fetch-mock'
import { ContentSvg } from 'pdfmake/interfaces'
import { eventQueryDataGenerator, User, UUID } from '@opencrvs/commons/client'
import { svgToPdfTemplate, stringifyEventMetadata } from './pdfUtils'

const fetch = createFetchMock(vi)
fetch.enableMocks()

const locations = [
  {
    id: '35391063-7dca-4e57-abd3-20dcc8538a64' as UUID,
    externalId: '2OKicPQMNI',
    name: 'HQ Office',
    partOf: 'f09c8dda-2156-420a-8215-2beda4c81d66' as UUID
  },
  {
    id: 'f09c8dda-2156-420a-8215-2beda4c81d66' as UUID,
    externalId: 'BxrIbNW7f3K',
    name: 'Embe',
    partOf: '7ef2b9c7-5e6d-49f6-ae05-656207d0fc64' as UUID
  },
  {
    id: '7ef2b9c7-5e6d-49f6-ae05-656207d0fc64' as UUID,
    externalId: 'B1u1bVtIA92',
    name: 'Pualula',
    partOf: null
  }
]
const userId = '677fb08730f3abfa33072769'

describe('stringifyEventMetadata', () => {
  test('Resolves event metadata', () => {
    const { declaration, ...metadata } = eventQueryDataGenerator({
      id: 'seabeast-clad-stad-elia-oleocellosis' as UUID,
      assignedTo: userId,
      createdByUserType: 'user',
      createdBy: userId,
      trackingId: 'B77FF6',
      createdAt: new Date(2000, 1, 1).toISOString(),
      updatedAt: new Date(2000, 1, 2).toISOString(),
      updatedAtLocation: locations[0].id,
      createdAtLocation: locations[0].id,
      updatedBy: userId
    })

    const users = [
      {
        id: userId,
        name: [
          {
            use: 'en',
            given: ['Joseph'],
            family: 'Musonda'
          }
        ],
        role: 'NATIONAL_REGISTRAR'
      }
    ] satisfies User[]

    const stringified = stringifyEventMetadata({
      metadata: {
        ...metadata,
        modifiedAt: new Date(2000, 1, 2).toISOString(),
        copiesPrintedForTemplate: 1
      },
      locations,
      users,
      intl: createIntl({ locale: 'en' })
    })
    expect(stringified).toMatchSnapshot()
  })
})
describe('svgToPdfTemplate', () => {
  test('replaces image URL with base64 data', async () => {
    fetch.mockResolvedValue({
      blob: async () =>
        Promise.resolve(new Blob(['fake-image-data'], { type: 'image/png' }))
    } as Response)

    const mockFiles = [
      'data:image/png;base64,FIRST_FILE_DATA',
      'data:image/png;base64,SECOND_FILE_DATA'
    ]

    global.FileReader = vi.fn(() => {
      const mockFileReader = {
        readAsDataURL: vi.fn(),
        result: mockFiles.shift(),
        onload: null as null | (() => void),
        onerror: null
      }
      // Trigger the FileReader onload manually
      setTimeout(() => {
        if (mockFileReader.onload) {
          mockFileReader.onload()
        }
      }, 0)
      return mockFileReader
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }) as any

    const svgString = `
      <svg width="100" height="100" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
        <image href="https://example.com/image.png" x="0" y="0" width="50" height="50"/>
        <image xlink:href="https://example.com/image2.png" x="0" y="0" width="50" height="50"/>
      </svg>
    `

    const result = await svgToPdfTemplate(svgString, {})
    const [content] = result.definition.content as [ContentSvg]

    expect(content).toHaveProperty('svg')
    expect(fetch).toHaveBeenCalledTimes(2)
    expect(content.svg).toBe(
      `
      <svg width="100" height="100" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
        <image href="data:image/png;base64,FIRST_FILE_DATA" x="0" y="0" width="50" height="50"/>
        <image xlink:href="data:image/png;base64,SECOND_FILE_DATA" x="0" y="0" width="50" height="50"/>
      </svg>
      `.trim()
    )
  })
})
