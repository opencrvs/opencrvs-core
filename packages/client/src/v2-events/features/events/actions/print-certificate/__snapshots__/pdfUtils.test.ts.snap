// Vitest Snapshot v1

exports[`stringifyEventMetadata > Resolves event metadata 1`] = `
{
  "assignedTo": {
    "name": "Joseph Musonda",
    "signature": "",
  },
  "createdAt": "1 February 2000",
  "createdAtLocation": {
    "country": "Farajaland",
    "district": "Embe",
    "location": "HQ Office",
    "province": "Pualula",
  },
  "createdBy": {
    "name": "Joseph Musonda",
    "signature": "",
  },
  "dateOfEvent": "1 February 2000",
  "flags": [],
  "id": "seabeast-clad-stad-elia-oleocellosis",
  "legalStatuses": {
    "DECLARED": null,
    "REGISTERED": null,
  },
  "modifiedAt": "2 February 2000",
  "status": "REGISTERED",
  "trackingId": "B77FF6",
  "type": "tennis-club-membership",
  "updatedAt": "2 February 2000",
  "updatedAtLocation": {
    "country": "Farajaland",
    "district": "Embe",
    "location": "HQ Office",
    "province": "Pualula",
  },
  "updatedBy": {
    "name": "Joseph Musonda",
    "signature": "",
  },
  "updatedByUserRole": "FIELD_AGENT",
}
`;
